### PR TITLE
Add --preservefds to podman run

### PR DESCRIPTION
**Pull request by QiWang19**
Add --preservefds to podman run. close https://github.com/containers/libpod/issues/6458

Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
/approve

**Comment by edsantiago** [comment]
Please add a system test, possibly in `test/system/030-run.bats`, using something like what I messaged you earlier today. The e2e test here is not very useful: all it does is confirm that the option does not cause a syntax error. It does not test any actual functionality.

File: cmd/podman/containers/run.go [hunk 1]
```diff
@@ -61,10 +61,12 @@ func runFlags(flags *pflag.FlagSet) {
 	flags.SetNormalizeFunc(common.AliasFlags)
 	flags.BoolVar(&runOpts.SigProxy, "sig-proxy", true, "Proxy received signals to the process")
 	flags.BoolVar(&runRmi, "rmi", false, "Remove container image unless used by other containers")
+	flags.UintVar(&runOpts.PreserveFDs, "preserve-fds", 0, "Pass N additional file descriptors to the container")
```
Comment (mheon) [inline cmd/podman/containers/run.go:64]: Recommend rewording to "Pass a number of additional file descriptors into the container"

File: libpod/container.go [hunk 2]
```diff
@@ -414,6 +414,11 @@ type ContainerConfig struct {
 
 	// HealthCheckConfig has the health check command and related timings
 	HealthCheckConfig *manifest.Schema2HealthConfig `json:"healthcheck"`
+
+	// PreserveFDs is a number of additional file descriptors (in addition
+	// to 0, 1, 2) that will be passed to the executed process. The total FDs
+	// passed will be 3 + PreserveFDs.
+	PreserveFDs uint `json:"-"`
```
Comment (mheon) [inline libpod/container.go:421]: Can you add a comment that this is not in the JSON because it is not supported for remote connections?

Comment (mheon) [reply libpod/container.go:421]: Oh, wait, nevermind - this needs to actually have a JSON tag here, not `"-"` - that should only be in Specgen. This is for the DB, and we do want to save this to the container in the DB.

File: libpod/options.go [hunk 3]
```diff
@@ -1369,6 +1369,17 @@ func WithHealthCheck(healthCheck *manifest.Schema2HealthConfig) CtrCreateOption
 	}
 }
 
+// WithPreserveFDs adds the file descriptors to the container config
```
Comment (mheon) [inline libpod/options.go:1372]: Suggest "WithPreserveFDs forwards from the process running Libpod into the container the given number of extra FDs (starting after the standard streams) to the created container"

File: pkg/specgen/specgen.go [hunk 4]
```diff
@@ -130,6 +130,10 @@ type ContainerBasicConfig struct {
 	// Remove indicates if the container should be removed once it has been started
 	// and exits
 	Remove bool `json:"remove"`
+	// PreserveFDs is a number of additional file descriptors (in addition
+	// to 0, 1, 2) that will be passed to the executed process. The total FDs
+	// passed will be 3 + PreserveFDs.
+	PreserveFDs uint `json:"-"`
```
Comment (mheon) [inline pkg/specgen/specgen.go:136]: We do need a comment here that this is JSON:"-" because it's not available remote.

Comment (mheon) [reply pkg/specgen/specgen.go:136]: Unlike container.go this one should stay "-" as well

File: docs/source/markdown/podman-run.1.md [hunk 5]
```diff
@@ -609,6 +609,10 @@ If a container is run with a pod, and the pod has an infra-container, the infra-
 
 Run container in an existing pod and read the pod's ID from the specified file.  If a container is run within a pod, and the pod has an infra-container, the infra-container will be started before the container is.
 
+**--preserve-fds**=*N*
+
+Pass down to the process N additional file descriptors (in addition to 0, 1, 2).  The total FDs will be 3+N.
```
Comment (mheon) [inline docs/source/markdown/podman-run.1.md:614]: Mention this is local-only

Comment (mheon) [reply docs/source/markdown/podman-run.1.md:614]: Actually, maybe not... I don't think we do this for --latest, do we?

Comment (QiWang19) [reply docs/source/markdown/podman-run.1.md:614]: > Actually, maybe not... I don't think we do this for --latest, do we?
>

 I don't see podman run has `--latest`. podman exec with both --latest and --preserve-fds seems working
```
$ podman exec -it --latest --preserve-fds=1 sh
/ # ls -l /proc/self/fd
total 0
lrwx------    1 root     root            64 Jun 16 19:52 0 -> /dev/pts/0
lrwx------    1 root     root            64 Jun 16 19:52 1 -> /dev/pts/0
lrwx------    1 root     root            64 Jun 16 19:52 2 -> /dev/pts/0
lrwx------    1 root     root            64 Jun 16 19:52 3 -> /null
ls: /proc/self/fd/4: cannot read link: No such file or directory
lr-x------    1 root     root            64 Jun 16 19:52 4
```


> Mention this is local-only

the `--preserve-fds` was hidden if remote in `podman exec`, so I also hide it for `podman run`. I can mention it is local-only.

Comment (mheon) [reply docs/source/markdown/podman-run.1.md:614]: I don't think we need further manpage changes, it looks like other flags hidden if remote are still in the manpages.

**Comment by QiWang19** [comment]
> Please add a system test, possibly in `test/system/030-run.bats`, using something like what I messaged you earlier today. The e2e test here is not very useful: all it does is confirm that the option does not cause a syntax error. It does not test any actual functionality.

https://github.com/containers/libpod/pull/6625/files#diff-f1c2f651579cf49df98160fa0e2d50a7 test failed. @edsantiago  Can you help me diagnose if it's because the test not correct? Otherwise, my implementation may not as expected.

**Comment by edsantiago** [comment]
@QiWang19 this is what the test should look like:
```bash
@test "podman run --preserve-fds" {
    skip_if_remote

    content=$(random_string 20)
    echo "$content" > $PODMAN_TMPDIR/tempfile

    run_podman run --rm -i --preserve-fds=2 $IMAGE sh -c "cat <&4" 4<$PODMAN_TMPDIR/tempfile
    is "$output" "$content" "container read input from fd 4"
}
```
Please use that. It still fails, but it does so for a real-problem reason which is being tracked in #6653 
```
 2020/06/17 15:51:44 bolt.Close(): funlock error: bad file descriptor
time="2020-06-17T15:51:44-06:00" level=error msg="failed to close libpod db: \"db file close: close /var/lib/containers/storage/libpod/bolt_state.db: bad file descriptor\""
Error: error saving container 26cd8352b73d3c07b46343c63f3b3b5c00e37f28ed685ace08788745025a0704 state: write /var/lib/containers/storage/libpod/bolt_state.db: bad file descriptor

```
Please coordinate efforts to resolve that issue.

Explanation of what I changed:
* use FD 4, not 3: I forgot that BATS uses fd 3 for internal purposes.
* Remove the NONLOCAL_IMAGE stuff. You copy-pasted that from another test. It is not necessary or desired here (it introduces a point of failure because it relies on remote registry)
* use random content, not hardcoded strings which could easily appear via other ways
* use a file in `$PODMAN_TMPDIR` - otherwise "testfile" will pollute your current working directory
* use `--rm` in the run command, to avoid having to clean up

**Comment by QiWang19** [comment]
@edsantiago Thanks!

**Comment by edsantiago** [comment]
@QiWang19 it looks like #6653 might not get attention soon. Since I believe your PR is correct, and the real problem is elsewhere, I'd be OK with you adding `skip "enable this once #6653 is fixed"` to the BATS test and trying to get your PR committed.

**Comment by mheon** [comment]
Hold this until Monday, 2.0 is feature-frozen

**Comment by rhatdan** [comment]
@mheon Are you ok to merge this now?

File: libpod/container.go [hunk 6]
```diff
@@ -418,6 +418,11 @@ type ContainerConfig struct {
 
 	// HealthCheckConfig has the health check command and related timings
 	HealthCheckConfig *manifest.Schema2HealthConfig `json:"healthcheck"`
+
+	// PreserveFDs is a number of additional file descriptors (in addition
+	// to 0, 1, 2) that will be passed to the executed process. The total FDs
+	// passed will be 3 + PreserveFDs.
+	PreserveFDs uint `json:"preserveFds,omitempty"`
```
Comment (mheon) [inline libpod/container.go:425]: This needs to be `json:"-"`

File: pkg/specgen/specgen.go (hunk 4 above)
Comment (mheon) [reply pkg/specgen/specgen.go:137]: This should not be `json:"-"`

File: libpod/container.go (hunk 6 above)
Comment (mheon) [reply libpod/container.go:425]: Wait, I'm confused, nevermind.

File: pkg/specgen/specgen.go (hunk 4 above)
Comment (mheon) [reply pkg/specgen/specgen.go:137]: Nevermind, I'm confused.

**Comment by mheon** [comment]
LGTM, but would like final signoff from @giuseppe

**Comment by QiWang19** [comment]
Thanks.
@giuseppe PTAL.

**Pull request merged**